Add hlint tool

jsha opened the pull request:
Hwæt! The key-holders in days of old have often held the wrong end of the stick. Strong chains of scrawled signatures tie anchors to entities, tie roots of trust to Toots about Rust. But bygone bindings blunder when bytewise basics are broken. Hlint, eagle-eyed and enterprising, seeks out SKIDs and Subjects, certifying that same is matched with same.

This checks a pile of certificates for consistent correspondence of Subject to subjectKeyIdentifier (SKID), and vice versa. If a single Subject appears with different SKIDs, or a single SKID appears with different Subjects, it produces an error.

Note that this is an opinionated tool: the is no general prohibition on distinct Subjects sharing a public key (and thus a SKID), but it is not our practice or intention to do so.

jsha committed:
Add hlint tool

Hwæt! The key-holders in days of old have often held the wrong end of the stick.
Strong chains of scrawled signatures tie anchors to entities, tie roots of trust
to Toots about Rust. But bygone bindings blunder when bytewise basics are
broken. Hlint, eagle-eyed and enterprising, seeks out SKIDs and Subjects,
certifying that same is matched with same.

This checks a pile of certificates for consistent correspondence of Subject to
subjectKeyIdentifier (SKID), and vice versa. If a single Subject appears with
different SKIDs, or a single SKID appears with different Subjects, it produces
an error.

Note that this is an opinionated tool: the is no general prohibition on distinct
Subjects sharing a public key (and thus a SKID), but it is not our practice or
intention to do so.

aarongable commented on hlint/README.md:
```diff
@@ -0,0 +1,16 @@
+Subjects sharing a public key (and thus a SKID), but it is not our practice or
+intention to do so.
+
+# Usage
```
```suggestion
## Usage
```

aarongable commented on hlint/main.go:
```diff
@@ -0,0 +1,75 @@
+	for _, c := range certs {
+		skid := hex.EncodeToString(c.SubjectKeyId)
+		if otherCert, ok := bySKID[skid]; ok && !bytes.Equal(c.RawSubject, otherCert.RawSubject) {
+			fmt.Fprintf(os.Stderr, "SKID %s has conflicting subjects in %s and %s: %s vs %s\n",
```
```suggestion
			fmt.Fprintf(os.Stderr, "SKID %s has conflicting subjects in %q and %q: %q vs %q\n",
```

aarongable commented on hlint/main.go:
```diff
@@ -0,0 +1,75 @@
+		}
+
+		if otherCert, ok := bySubject[string(c.RawSubject)]; ok && !bytes.Equal(c.SubjectKeyId, otherCert.SubjectKeyId) {
+			fmt.Fprintf(os.Stderr, "subject %s has conflicting SKIDs in %s and %s: %x vs %x\n",
```
```suggestion
			fmt.Fprintf(os.Stderr, "subject %q has conflicting SKIDs in %q and %q: %x vs %x\n",
```

aarongable commented on hlint/main.go:
```diff
@@ -0,0 +1,75 @@
+}
+
+func readCert(filename string) (*certificate, error) {
+	bytes, err := ioutil.ReadFile(filename)
```
```suggestion
	bytes, err := os.ReadFile(filename)
```

aarongable reviewed (commented):
A few tiny nits, and a suggestion: add this to one or more of the run.sh files now, to "show" it working.

jsha committed:
Review feedback

jsha committed:
Invoke hlint in 2024 ceremony

jsha commented:
Updated, thanks! I added it to only the 2024 run.sh.

Worth noting that if I run this against the current outputs directory it does reveal two inconsistencies:

```
/tmp/ceremony-tools/bin/hlint outputs/*/*.cert.pem
SKID 39f19db10e168971f7f8dc2a55dc09cc835a78bb has conflicting subjects in "outputs/2020/root-x2.cert.pem" and "outputs/2020/root-x2-cross.cert.pem": "CN=(FAKE) ISRG Root X2,O=Internet Security Research Group,C=US" vs "CN=(FAKE) ISRG Root X2,O=(FAKE) Internet Security Research Group,C=US"
SKID 66acf1bc77311678eb3aed855c4d1ef5972ab925 has conflicting subjects in "outputs/2021/root-x1-cross.cert.pem" and "outputs/2015/root-x1.cert.pem": "CN=(FAKE) ISRG Root X1,O=(FAKE) Internet Security Research Group,C=US" vs "CN=(FAKE) ISRG Root X1,O=Internet Security Research Group,C=US"
```

aarongable commented on ceremonies/2024/run.sh:
```diff
@@ -109,3 +109,9 @@ openssl verify \
+GOBIN="${TOOLS}/bin" go install ./
+cd -
+
+"${TOOLS}/bin/hlint" *.cert.pem
```
Looks like this last line is missing a \n

aarongable commented on ceremonies/2024/run.sh:
```diff
@@ -109,3 +109,9 @@ openssl verify \
     "./int-e8.cert.pem" \
     "./int-e9.cert.pem"
+
+cd ../../hlint
+GOBIN="${TOOLS}/bin" go install ./
+cd -
```
This install feels like it should happen inside setup_ceremony_tool

jsha replied:
I wound up putting this in its own setup function. `setup_ceremony_tool` is gated by an existence check for an executable at `${CEREMONY_BIN_2024}`, and it feels not really correct to have the creation of two executables gated on the existence of just one of them (and gating on the existence of both also introduces weird cases).

aarongable reviewed (commented):
LGTM with two nits. Honestly I'd consider fixing the inconsistency that it found already; we're under no obligation to leave the older configs untouched -- they're a tool for us and if they're not optimally useful we should fix them.

jsha committed:
add setup function for hlint

jsha commented:
> I'd consider fixing the inconsistency that it found already

Actually it looks like you already fixed that inconsistency in #17 and updated the configs 🎉

I suspect that was part of the same "huh why doesn't this work" investigation that led us to conclude something like hlint would be useful. 🔄
